akka-cluster/src/test/scala/akka/cluster/JoinConfigCompatCheckerSpec.scala

```diff
@@ -0,0 +1,359 @@
+package akka.cluster
+
+import akka.actor.{ ActorSystem, Address }
+import akka.cluster.InternalClusterAction.LeaderActionsTick
+import akka.cluster.MemberStatus.{ Removed, Up }
+import akka.testkit.{ AkkaSpec, LongRunningTest }
+import com.typesafe.config.{ Config, ConfigFactory }
+import org.scalatest.{ Matchers, WordSpec }
+
+import scala.concurrent.duration._
+import scala.collection.JavaConverters._
+import scala.collection.{ immutable ⇒ im }
+
+object JoinConfigCompatCheckerSpec {
+
+}
+
+class JoinConfigCompatCheckerSpec extends AkkaSpec() {
```

Comment:
@patriknw, 

I'm not 100% satisfied with this test. There are many cases that we can't easily assert because it's all happening behind de scenes. 

For instance, when a node fails to join a cluster, we can't be sure of the real cause. Did the cluster send an `IncompatibelConfg` message back or did it send `CompatibelConfg`, but the cluster config wasn't successfully validated on the joining side?

Same with the `ensure-on-join` flag. Once enabled, we don't know if the node joins because it ignored the validation or because it simply passed. 

And there is also the `FirstSeedNodeProcess` that are much harder to test because we need to first create a cluster with two nodes, remove the first, update its config and try to make it re-join the cluster using the same APIs address.